### PR TITLE
Add functionality to compare DataFrames

### PR DESCRIPTION
**Pull request by alysivji**
Added functionality to compare DataFrames by leveraging ```pandas.util.testing``` assert helpers

### CODE REVIEW

File: engarde/checks.py [hunk 1]
```diff
@@ -246,7 +247,28 @@ def one_to_many(df, unitcol, manycol):
     return df
 
 
-__all__ = ['is_monotonic', 'is_shape', 'none_missing', 'unique_index', 'within_n_std',
-           'within_range', 'within_set', 'has_dtypes',
-           'verify', 'verify_all', 'verify_any']
+def is_same_as(df, df_to_compare):
```
Comment (TomAugspurger) [inline engarde/checks.py]: Can you accept `**kwargs` here and pass them through to the `tm.assert_frame_equal`? That can be useful for relaxing the equality requirements.

Comment (alysivji) [inline engarde/checks.py]: How should I acknowledge ```kwargs``` in the docstring?

Comment (TomAugspurger) [inline engarde/checks.py]: Something like

```
**kwargs : dict
    keyword arguments passed through to ``assert_frame_equal``
```

File: engarde/checks.py [hunk 2]
```diff
@@ -246,7 +247,30 @@ def one_to_many(df, unitcol, manycol):
...
+    **kwargs : dict
+        keyword arguments passed through to panda's ``assert_frame_equal``
+
+    Returns
+    =======
+    df : DataFrame
+
+    """
+    try:
+        tm.assert_frame_equal(df, df_to_compare)
```
Comment (TomAugspurger) [inline engarde/checks.py]: You'll have to pass through `**kwargs` here to `tm.assert_frame_equal` :) This lets you do things like

```python
df.pipe(is_same_as, other, check_names=False)
```

File: engarde/checks.py [hunk 3]
```diff
@@ -246,7 +247,30 @@ def one_to_many(df, unitcol, manycol):
...
+    except AssertionError as e:
+        raise AssertionError("DataFrames are not equal")
```
Comment (TomAugspurger) [inline engarde/checks.py]: It'd be nice to include the original assertion error for why they aren't equal here. I know in python3 it's just

```python
raise AseertionError("DataFrames are not equal") from e
```

but I don't know about py2. I think leave it for now.

Are you using python3 or 2? I'm tempted to just make the next release python3 only.

File: engarde/decorators.py [hunk 4]
```diff
@@ -161,7 +161,19 @@ def wrapper(*operation_args, **operation_kwargs):
...
     return decorate
 
-__all__ = ['is_monotonic', 'is_shape', 'none_missing', 'unique_index',
-           'within_range', 'within_set', 'has_dtypes', 'verify', 'verify_all',
-           'verify_any', 'within_n_std']
+
+def is_same_as(df_to_compare, **kwargs):
+    def decorate(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
```
Comment (TomAugspurger) [inline engarde/decorators.py]: Ugh, this is a bit complicated. Can you rename the outer `is_same_as` `**kwargs` to `**assert_kwargs` and then pass those to `ck.is_same_as` on line 170? Then the `wrapper` `**kwargs` are for whatever `func` is being decorated (like you have now).

**Comment by alysivji** [comment]
I got the other errors fixed with tests around **kwargs functionality to ensure that less strict DataFrames match (float vs int).

I dug into re-raising exceptions and found [this post on StackOverflow](https://stackoverflow.com/questions/696047/re-raise-exception-with-a-different-type-and-message-preserving-existing-inform/792163#792163). Reading thru the referenced blogs ([1](http://www.ianbicking.org/blog/2007/09/re-raising-exceptions.html), [2](https://nedbatchelder.com/blog/200711/rethrowing_exceptions_in_python.html)), we can provide better tracebacks using a three-argument raise statement for python2.

Personally, I'm using Python3 (#Teamfstrings). I do not have much experience in developing for both. Would it be difficult to implement two different exception re-raising workflows? How do other libraries handle this?

There are definitely good reasons to keep testing libraries like this compatible with both py2 and py3. I can help out if you decide to go this route.

For now, I amended my last commit to fix **kwargs usage, added better tests around that functionality, and kept the ```raise AssertionError("DataFrames are not equal")``` functionality the same.

**Comment by TomAugspurger** [comment]
This looks good. I'll figure out the py2 vs. py3 stuff later.

Hopefully will do a release in a couple weeks.

**Comment by alysivji** [comment]
I'm listening to Hynek's talk from PyCon while I'm working. He mentioned the [```raise_from()``` method in the six library](https://pythonhosted.org/six/#six.raise_from).

This should fix all exception chaining issues. I'll update my code when I get some time this weekend.

**Pull request merged**